src/com/xilinx/rapidwright/design/tools/NetlistClockDetection.java

```diff
@@ -0,0 +1,180 @@
+        versalPrimsToClockPins.put("MBUFGCE_DIV", Stream.of("I").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("MBUFGCE", Stream.of("I").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("MBUFGCTRL", Stream.of("I0", "I1").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("MBUFG_GT", Stream.of("I").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("MBUFG_PS", Stream.of("I").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("MMCME5", Stream.of("DCLK", "PSCLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("ODDRE1", Stream.of("C").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("ODELAYE5", Stream.of("CLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("XPLL", Stream.of("DCLK", "PSCLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("RAM32X16DR8", Stream.of("WCLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("RAM64X8SW", Stream.of("WCLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("RAMB18E5_INT", Stream.of("CLKARDCLK", "CLKBWRCLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("RAMB36E5_INT", Stream.of("CLKARDCLKL", "CLKARDCLKU", "CLKBWRCLKL", "CLKBWRCLKU").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("SRL16E", Stream.of("CLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("URAM288E5_BASE", Stream.of("CLK").collect(Collectors.toSet()));
+        versalPrimsToClockPins.put("URAM288E5", Stream.of("CLK").collect(Collectors.toSet()));
+        seriesPrimsToClockPins.put(Series.Versal, versalPrimsToClockPins);
+    }
+
+    /**
+     * Finds all terminal hierarchical port insts (pins) which gate changes to the given
+     * hierarchical net's value, i.e., finds all terminal pins which may be drivers of
+     * the net through non-clocked primitives and all likely terminal clock signal pins
+     * that gate signals which drive the net through clocked primitives.
+     *
+     * For example, when run on the output hierarchical net of a LUT2 whose inputs come
+     * from a flop and a LUT, this will return the terminal hierarchical port inst(s)
+     * driving the flop's clock pin as well as any gating drivers of the input LUT's
+     * inputs, determined recursively.
+     *
+     * @param hNet             The hierarchical net to find gating drivers of.
+     * @param primsToClockPins A map from primitive names to their input clock pins, if
+     *                         the primitive does have input clock pins. This is
+     *                         specific to the FPGA series.
+     * @param encountered      A set of already-encountered EDIFHierNets to skip during
+     *                         netlist traversal.
+     */
+    static Set<EDIFHierPortInst> getGatingDrivers(EDIFHierNet hNet, Map<String, Set<String>> primsToClockPins, Set<EDIFHierNet> encountered) {
```

Comment:
```suggestion
    private static Set<EDIFHierPortInst> getGatingDrivers(EDIFHierNet hNet, Map<String, Set<String>> primsToClockPins, Set<EDIFHierNet> encountered) {
```
I'm assuming we don't want outside users calling this one?